### PR TITLE
Add first OCI runtime abstraction

### PR DESCRIPTION
**Pull request by saschagrunert**
#### What type of PR is this?

/kind design

#### What this PR does / why we need it:
This is the first OCI runtime trait to be used to implement runtimes
like runc/crun.
#### Which issue(s) this PR fixes:

None
#### Special notes for your reviewer:
@mrunalp I would appreciate any feedback if this design will fit our needs.
#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

File: src/oci/runtime/runc.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    /// Retrieve the state of a container.
+    async fn state(&self, _container: &RuncContainer) -> Result<ContainerState> {
+        unimplemented!()
+    }
+}
+
+#[derive(Debug, Default, Builder, Getters)]
+#[builder(default, pattern = "owned", setter(into, strip_option))]
+/// A runc container abstraction.
+pub struct RuncContainer {
```
Comment (mrunalp) [inline src/oci/runtime/runc.rs:79]: We can call this `OCIContainer` or similar and support `runc` and `crun` together :)

Comment (saschagrunert) [reply src/oci/runtime/runc.rs]: Sounds good! Do you think we even need the associated type on the trait? We could also make `Container` a `struct` (not a `trait`) and pre-define the data we need, like the `spec`. Not sure if this decreases the flexibility.

I also renamed the `Runc` implementation to `Cli`. That's more generic and we can tell the user that this implementation is for CLI-compatible OCI runtimes, like runc and crun.

File: src/oci/runtime/cli.rs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+use crate::oci::runtime::{Container, ContainerState, ContainerStats, Runtime};
+use crate::oci::spec::runtime::{LinuxResources, Spec};
+use anyhow::Result;
+use async_trait::async_trait;
+use derive_builder::Builder;
+use getset::Getters;
+use tokio::{process::Command, signal::unix::SignalKind};
+
+struct Cli;
```
Comment (mrunalp) [inline src/oci/runtime/cli.rs:15]: I think we should call this something else.

File: src/oci/runtime/cli.rs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    async fn state(&self, _container: &OCIContainer) -> Result<ContainerState> {
+        unimplemented!()
+    }
+}
+
+#[derive(Debug, Default, Builder, Getters)]
+#[builder(default, pattern = "owned", setter(into, strip_option))]
+/// A general container abstraction.
+pub struct OCIContainer {
+    #[get]
```
Comment (mrunalp) [inline src/oci/runtime/cli.rs:84]: We need an id as well at minimum.

File: src/oci/runtime/mod.rs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+    /// Retrieve container resource statistics.
+    async fn stats(&self, container: &Self::Container) -> Result<ContainerStats>;
+
+    /// Retrieve the state of a container.
+    async fn state(&self, container: &Self::Container) -> Result<ContainerState>;
+}
+
+/// Container is the implementation of an OCI compatible container.
+pub trait Container {}
```
Comment (mrunalp) [inline src/oci/runtime/mod.rs:56]: Hmm, I am not sure why we need this.

File: src/oci/runtime/mod.rs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+use anyhow::Result;
+use async_trait::async_trait;
+use strum::{AsRefStr, Display, EnumString, IntoStaticStr};
+use tokio::{process::Command, signal::unix::SignalKind};
+
+pub mod cli;
+
+#[async_trait]
+/// Runtime is the trait for implementing possible interactions with an OCI compatible container.
+pub trait Runtime {
```
Comment (mrunalp) [inline src/oci/runtime/mod.rs:13]: I think that this trait is more appropriately named as `Container` :)

File: src/oci/runtime/runc.rs (hunk 1 above)
Comment (mrunalp) [reply src/oci/runtime/runc.rs]: I think we can use the associate type :+1: and maybe aggregate the concrete ones with a `BaseContainer` type that has information common to all container types.

File: src/oci/runtime/mod.rs (hunk 5 above)
Comment (saschagrunert) [reply src/oci/runtime/mod.rs:13]: Yep, renamed it. :+1:

File: src/oci/runtime/mod.rs (hunk 4 above)
Comment (saschagrunert) [reply src/oci/runtime/mod.rs:56]: Removed this trait because `Runtime` is now `Container`.

File: src/oci/runtime/cli.rs (hunk 2 above)
Comment (saschagrunert) [reply src/oci/runtime/cli.rs]: I renamed it to `local`.

File: src/oci/runtime/cli.rs (hunk 3 above)
Comment (saschagrunert) [reply src/oci/runtime/cli.rs]: Added an `id` field + docs. :+1:

File: src/oci/runtime/runc.rs (hunk 1 above)
Comment (saschagrunert) [reply src/oci/runtime/runc.rs]: With the renaming of `Runtime` to `Container` we can directly refer to `Self` as type. Now I do not see any reason for the associated type any more.

The `create` method is gone in replacement of the builder pattern, too.

File: src/oci/container/local.rs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    id: String,
+
+    #[get = "pub"]
+    /// OCI Runtime Specification of the container.
+    spec: Spec,
+}
+
+#[async_trait]
+impl Container for OCIContainer {
+    /// Execute the user defined process in a created container.
```
Comment (mrunalp) [inline src/oci/container/local.rs:33]: we are missing `create`

Comment (saschagrunert) [reply src/oci/container/local.rs:38]: Yep, added the method :+1:

**Review by mrunalp — APPROVED** [review]

**Pull request merged**